Add the dyn logDir support

lxbsz opened the pull request:
Currently we have support the dyn-config for the lruCount and logLevel, this PR will add the logDir support.
It will also change the logrotate's config file at the same time.

Thanks.

lxbsz commented:
@pkalever @amarts 
Please review, thanks.

pkalever commented on utils/lru.c:
```diff
@@ -22,6 +24,31 @@ typedef struct Entry {
+    tmp = list_entry(pos, Entry, list);
+    if (glfs_set_logging(tmp->glfs, logPath, GFAPI_LOG_LEVEL)) {
+      LOG("mgmt", GB_LOG_WARNING, "glfs_set_logging(%s, %d) on %s failed[%s]",
+          logPath, GFAPI_LOG_LEVEL, tmp->volume, strerror(errno));
```
May be later, it make more sense to covert GFAPI_LOG_LEVEL as configurable.

lxbsz replied:
Yeah, If you agree, let's do it in late future.
Thanks.

pkalever commented on utils/utils.c:
```diff
@@ -278,24 +280,43 @@ void fetchGlfsVolServerFromEnv()
   if (strlen(logDir) > PATH_MAX - GB_MAX_LOGFILENAME) {
     fprintf(stderr, "strlen of logDir Path > PATH_MAX: %s", logDir);
     return EXIT_FAILURE;
+    goto out;
```
return and goto ? My be you want to assign  ret = EXIT_FAILURE; and goto out;

lxbsz replied:
Yeah, will fix it.
Thanks.

pkalever commented on utils/dyn-config.c:
```diff
@@ -167,11 +167,20 @@ glusterBlockGetOption(const char *key)
 
 static void
-glusterBlockConfSetOptions(gbConfig *cfg, bool reloading)
+glusterBlockConfSetOptions(gbConfig *cfg, bool getLogDir)
```
May be you want to change the same in the declaration too, in common.h ?
```

- utils/common.h:int glusterBlockLoadConfig(gbConfig *cfg, bool reloading);
+ utils/common.h:int glusterBlockLoadConfig(gbConfig *cfg, bool getLogDir);
```

lxbsz replied:
Will update it.
Thanks.

pkalever commented on cli/gluster-block.c:
```diff
@@ -1037,7 +1037,7 @@ main(int argc, char *argv[])
   }
 
-  if(initLogging()) {
+  if(initLogging("cli")) {
```
Can we use:
gbProcessCtx gbCtx = GB_CLI_MODE; /* set process mode */   ?

lxbsz replied:
WIll fix it.
Thanks.

pkalever commented on daemon/gluster-blockd.c:
```diff
@@ -587,7 +587,7 @@ main (int argc, char **argv)
 
 
-  if(initLogging()) {
+  if(initLogging("mgmt")) {
```
Same here:
 gbProcessCtx gbCtx = GB_DAEMON_MODE; /* set process mode */ ?

Intead of passing the argument ?

lxbsz replied:
Will fix it.
Thanks.

pkalever commented on utils/utils.h:
```diff
@@ -637,7 +638,7 @@ void fetchGlfsVolServerFromEnv(void);
 bool glusterBlockSetLogDir(char *logDir);
 
-int initLogging(void);
+int initLogging(const char *dom);
```
My initial comment about gbCtx also applies here.

lxbsz replied:
Yeah, removed.
Thanks.

pkalever commented on utils/dyn-config.c:
```diff
@@ -465,12 +475,53 @@ glusterBlockLoadConfig(gbConfig *cfg, bool reloading)
+
+
+  if (GB_ALLOC(cfg) < 0) {
+    MSG(stderr, "Alloc GB config failed for configPath: %s!\n", configPath);
```
With the recent fixes '\n' is no more needed here.
Please fix this else where too.

lxbsz replied:
Will fix it.
Thanks.

pkalever commented on utils/utils.c:
```diff
@@ -280,12 +281,87 @@ void fetchGlfsVolServerFromEnv()
+  fp = fopen(GB_LOGROTATE_PATH, "r+");
+  if (fp == NULL) {
+    ret = -errno;
+    LOG(dom, GB_LOG_ERROR, "Failed to open file '%s', %m\n", GB_LOGROTATE_PATH);
```
No '\n' needed here and other places too.

lxbsz replied:
Will fix it.
Thanks.

pkalever commented on utils/utils.c:
```diff
@@ -280,12 +281,87 @@ void fetchGlfsVolServerFromEnv()
+  if (ret != len) {
+    LOG(dom, GB_LOG_ERROR, "Failed to update '%s', %m\n", GB_LOGROTATE_PATH);
+    goto error;
+  }
```
I have not yet tested, but I assume this will not override the first line:
```
$ cat /etc/logrotate.d/gluster-block
# Rotate all the gluster-block logs                                              
/var/log/gluster-block/*.log {         
```

lxbsz replied:
Yeah, it won't. I will only replace the second line above.
Thanks.

pkalever commented:
@lxbsz One generic question here:
After we update the /etc/logrotate.d/gluster-block file with the new log location, how long does it take for log-rotate to pick this ? is this new change dynamically reloaded with in logrotate ?

Thanks!

pkalever commented on utils/utils.c:
```diff
@@ -280,14 +281,111 @@ void fetchGlfsVolServerFromEnv()
+      dom = "cli";
+  } else if (gbCtx == GB_DAEMON_MODE) {
+      dom = "mgmt";
+  }
```
@lxbsz 
code missing alignment of 2 spaces here.

lxbsz replied:
Will fix it.

pkalever commented on utils/utils.c:
```diff
@@ -280,14 +281,111 @@ void fetchGlfsVolServerFromEnv()
+  fp = fopen(GB_LOGROTATE_PATH, "r+");
+  if (fp == NULL) {
+    ret = -errno;
+    LOG(dom, GB_LOG_ERROR, "Failed to open file '%s', %m", GB_LOGROTATE_PATH);
```
@lxbsz One more thing:

instead of %m, use stderror(errno)

I remember there was a patch to remove all such uses, may be a co-verity fix.

lxbsz replied:
Will fix it.
Thanks.

pkalever commented:
@lxbsz Noted the compile time warning:

```
  CC       libgb_la-dyn-config.lo                                                                                     
In file included from common.h:15,                                                                                    
                 from lru.h:17,                                                                                       
                 from lru.c:12:                                                                                       
lru.c: In function ‘glusterBlockSetLruCount’:              
utils.h:114:27: warning: implicit declaration of function ‘asprintf’; did you mean ‘vsprintf’? [-Wimplicit-function-de
claration]                                                 
                 int __ret=asprintf(ptr, ##fmt);__ret;})   
                           ^~~~~~~~                        
utils.h:132:19: note: in expansion of macro ‘GB_ASPRINTF’  
             len = GB_ASPRINTF(&buf, fmt, ##__VA_ARGS__);             \                                               
                   ^~~~~~~~~~~                             
lru.c:56:5: note: in expansion of macro ‘MSG’              
     MSG(stderr, "glfsLruCount should be [0 < COUNT < %d]",
     ^~~

```

lxbsz commented on systemd/gluster-blockd.sysconfig:
```diff
@@ -23,5 +22,9 @@
 #GB_CLI_TIMEOUT=300
 
+# Supported setting the log directory path from the sysconfig
+# default path is /var/log/gluster-blockd, uncomment it and set your path:
```
Will fix it.

pkalever commented:
@lxbsz on Fedora29 when I tested with this patch, the dynconfig is not working as expected.

I did below:
On first change of LOG_LEVEL=DEBUG and :wq!, no change in loglevel
On second time open and :wq!, no change
ON third time open and :wq!, this time I saw
[2019-04-25 11:30:00.414131] CRIT: logLevel now is DEBUG [at utils.c+57 :<glusterBlockSetLogLevel>]                   
[2019-04-25 11:30:00.414315] DEBUG: event->mask: 0x2 [at dyn-config.c+573 :<glusterBlockDynConfigStart>] 


Can you have a check, thanks!

pkalever commented on utils/utils.c:
```diff
@@ -278,24 +280,43 @@ void fetchGlfsVolServerFromEnv()
-int
-initLogging(void)
+static int
+initLogDirAndFiles(char *newLogDir)
```
@lxbsz 
Where is that "CRIT: logDir now is /var/log/gluster-block' message :-)

lxbsz replied:
Will add this :-)
Thanks.

lxbsz commented:
> @lxbsz One generic question here:
> After we update the /etc/logrotate.d/gluster-block file with the new log location, how long does it take for log-rotate to pick this ? is this new change dynamically reloaded with in logrotate ?

Yeah, that's right.
Actually the logrotate service will read the glsuter-block logrotate config file when the timer is fired.

> 
> Thanks!

Updated, please review.

My test output:

```
[~]# cat /etc/logrotate.d/gluster-block
# Rotate all the gluster-block logs
/var/log/gluster-block/*.log {
  sharedscripts
  weekly
  maxsize 10M
  minsize 100k

# 6 months of logs are good enough
  rotate 26

  missingok
  compress
  delaycompress
  notifempty
  postrotate
  killall -q -s 1 gluster-blockd > /dev/null 2>&1 || true
  endscript
}
[~]# cat /etc/logrotate.d/gluster-block
# Rotate all the gluster-block logs
/var/log/gluster-block/tmp/aaa/*.log {
  sharedscripts
  weekly
  maxsize 10M
  minsize 100k

# 6 months of logs are good enough
  rotate 26

  missingok
  compress
  delaycompress
  notifempty
  postrotate
  killall -q -s 1 gluster-blockd > /dev/null 2>&1 || true
  endscript
}
[~]# 

```

> @lxbsz on Fedora29 when I tested with this patch, the dynconfig is not working as expected.
> 
> I did below:
> On first change of LOG_LEVEL=DEBUG and :wq!, no change in loglevel
> On second time open and :wq!, no change
> ON third time open and :wq!, this time I saw
> [2019-04-25 11:30:00.414131] CRIT: logLevel now is DEBUG [at utils.c+57 :]
> [2019-04-25 11:30:00.414315] DEBUG: event->mask: 0x2 [at dyn-config.c+573 :]
> 
> Can you have a check, thanks!

I am also hitting this issue, it seems due to adding another separate process to handle the TIRPC issue.
That means there will be 2 threads will watch the sysconfig dir at the same time.

Not sure whether will it be a problem.

Any idea ?

lxbsz commented:
> @lxbsz on Fedora29 when I tested with this patch, the dynconfig is not working as expected.
> 
> I did below:
> On first change of LOG_LEVEL=DEBUG and :wq!, no change in loglevel
> On second time open and :wq!, no change
> ON third time open and :wq!, this time I saw
> [2019-04-25 11:30:00.414131] CRIT: logLevel now is DEBUG [at utils.c+57 :]
> [2019-04-25 11:30:00.414315] DEBUG: event->mask: 0x2 [at dyn-config.c+573 :]
> 
> Can you have a check, thanks!

And please be aware that there will also be some problems for the LOG() and others after the TIRPC PRs, It is possible hat the 2 processes will update the same log files at the same time. And the gbConf.lock won't work as before.

lxbsz commented:
@pkalever 
Update all, please review again.
- There has one deadlock : gbConf.lock
- Fixed directory creation issue by using gbRunner("mkdir -p ...").
- Added the logs that shows the logDir's changes.
- Other fixing mentioned above.

Thanks.

pkalever commented:
Adding our discussion for others:

- Lets use some shared memory b/w parent (cli process) and child (server process) to share gbConf variable. Else as the parent is now running dynconfig thread, server process will not get updates about changes to configuration with /etc/sysconfig/gluster-blockd

How doing it: http://www.csl.mtu.edu/cs4411.ck/www/NOTES/process/shm/what-is-shm.html

- @lxbsz to identify if there are any other context/maintenance variables that needs to be shared b/w processes.

Thanks!

lxbsz commented:
> Adding our discussion for others:
> 
> * Lets use some shared memory b/w parent (cli process) and child (server process) to share gbConf variable. Else as the parent is now running dynconfig thread, server process will not get updates about changes to configuration with /etc/sysconfig/gluster-blockd
> 
That' right.

> How doing it: http://www.csl.mtu.edu/cs4411.ck/www/NOTES/process/shm/what-is-shm.html

After this in the child process(server process) it will only read the gbConf and the parent process will update it and read it.

> 
> * @lxbsz to identify if there are any other context/maintenance variables that needs to be shared b/w processes.

Will do it and will raise one new PR to address this issue.

Thanks.
> 
> Thanks!

pkalever commented on utils/utils.c:
```diff
@@ -283,14 +284,118 @@ void fetchGlfsVolServerFromEnv()
+    logLevel = GB_LOG_DEBUG;
+  } else if (gbCtx == GB_DAEMON_MODE) {
+    dom = "mgmt";
+    logLevel = GB_LOG_CRIT;
```
@lxbsz why setting logLevel here ?

pkalever replied:
ignore, I got it.

pkalever replied:
@lxbsz doesn't it makesense to log about LOGDIR change in CRIT level for cli too ?

lxbsz replied:
@pkalever 
IMO the cli is one short process in a very short time, it's okay to keep it as DEBUG.
Make sense ?

lxbsz replied:
If you want to change it to CRIT, I can update it right now.

pkalever replied:
But then, if we do that, I think on every cli call we will log it :-D

Lets live with it for now, no changes. Thanks!

pkalever commented:
> > Adding our discussion for others:
> > 
> > * Lets use some shared memory b/w parent (cli process) and child (server process) to share gbConf variable. Else as the parent is now running dynconfig thread, server process will not get updates about changes to configuration with /etc/sysconfig/gluster-blockd
> 
> That' right.
> 
> > How doing it: http://www.csl.mtu.edu/cs4411.ck/www/NOTES/process/shm/what-is-shm.html
> 
> After this in the child process(server process) it will only read the gbConf and the parent process will update it and read it.
> 
> > * @lxbsz to identify if there are any other context/maintenance variables that needs to be shared b/w processes.
> 
> Will do it and will raise one new PR to address this issue.
> 
> Thanks.
> 
> > Thanks!

lxbsz commented:
@pkalever 
Update the tag.
Thanks.

pkalever commented:
Thanks @lxbsz
